File: src/main/java/com/jb/Project_coupon_3/models/Coupon.java
package com.jb.Project_coupon_3.models;

import com.fasterxml.jackson.annotation.JsonGetter;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.sql.Date;
import java.util.Objects;
import java.util.Set;

@Entity
@Table(name = "coupons")
@Data
@NoArgsConstructor
public class Coupon {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Setter(AccessLevel.NONE)
    private int id;

    @ManyToOne(fetch = FetchType.EAGER)
//    @JsonIgnore
    private Company company;

    @Enumerated(value = EnumType.STRING)
    private Category category;

    @Column(nullable = false)
    private String title;

    private String description;

    @Column(nullable = false)
    private Date startDate;

    @Column(nullable = false)
    private Date endDate;

    private int amount;

    @Column(nullable = false)
    private double price;

    @Column(columnDefinition="LONGTEXT")
    private String image;

    @JsonIgnore
    @ManyToMany( fetch = FetchType.EAGER)
    Set<Customer> customers;

    public Coupon(Company company, Category category, String title, String description, Date startDate, Date endDate, int amount, double price, String image) {
        this.company = company;
        this.category = category;
        this.title = title;
        this.description = description;
        this.startDate = startDate;
        this.endDate = endDate;
        this.amount = amount;
        this.price = price;
        this.image = image;
//       this.customers = customers;
    }



    @Override
    public String toString() {
        return " \nCoupon{" +
                "id= " + id +
                "company=" + this.company.getId() +
                ", category=" + category +
                ", title='" + title + '\'' +
                ", description='" + description + '\'' +
                ", startDate=" + startDate +
                ", endDate=" + endDate +
                ", amount=" + amount +
                ", price=" + price +
                ", image='" + image +  '}';
    }

    @JsonGetter("company")
    public int companyId(){
        return company.getId();
    }


    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Coupon coupon = (Coupon) o;
        return id == coupon.id && amount == coupon.amount && Double.compare(price, coupon.price) == 0 && category == coupon.category && Objects.equals(title, coupon.title) && Objects.equals(description, coupon.description) && Objects.equals(startDate, coupon.startDate) && Objects.equals(endDate, coupon.endDate);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, category, title, description, startDate, endDate, amount, price);
    }

}

File: coupons3/src/Models/Coupon.ts
import {Category} from "./Category";
import Company from "./Company";

class Coupon {
    public title: string;
    public description: string;
    public startDate: Date;
    public endDate: Date;
    public category: Category;
    public amount: number;
    public price: number;
    public image: any;
    public company?:Company;
    public id?: number;


    constructor(title: string, description: string, startDate: Date, endDate: Date, category: Category, amount: number, price: number, image: any,company?: Company, id?: number) {
        this.company = company;
        this.title = title;
        this.description = description;
        this.startDate = startDate;
        this.endDate = endDate;
        this.category = category;
        this.amount = amount;
        this.price = price;
        this.image = image;
        this.id = id;
    }

// constructor(id: number,  title: string, description: string, startDate: Date, endDate: Date,category: Category, amount: number, price: number, image: string) {
    //     this._id = id;
    //     this._company = new Company(companyId);
    //     this._title = title;
    //     this._description = description;
    //     this._startDate = startDate;
    //     this._endDate = endDate;
    //     this._category = category;
    //     this._amount = amount;
    //     this._price = price;
    //     this._image = image;
    // }


    // toJSON(): any{
    //     return {
    //         "company":this._company ,
    //         "category": this._category,
    //         "title": this._title,
    //         "description": this._description,
    //         "startDate": this._startDate,
    //         "endDate": this._endDate,
    //         "amount": this._amount,
    //         "price": this._price,
    //         "image": this._image
    //     }
    // }



}

export default Coupon